lib/autoconfig.js

```diff
@@ -0,0 +1,289 @@
+ */
+function combineArrays(arr1, arr2) {
+    var res = [];
+    if (arr1.length === 0) {
+        return explodeArray(arr2);
+    }
+    if (arr2.length === 0) {
+        return explodeArray(arr1);
+    }
+    arr1.forEach(function(x1) {
+        arr2.forEach(function(x2) {
+            res.push([].concat(x1, x2));
+        });
+    });
+    return res;
+}
+
+//------------------------------------------------------------------------------
+// Private
+//------------------------------------------------------------------------------
+
+/**
+ * Group together valid rule configurations based on object properties
+ *
+ * e.g.:
+ * groupByProperty([
+ *     {beofre: true},
+ *     {before: false},
+ *     {after: true},
+ *     {after: false}
+ * ]);
+ *
+ * will return:
+ * [
+ *     [{before: true}, {before: false}],
+ *     [{after: true}, {after: false}]
+ * ]
+ *
+ * @param   {Object[]} objects Array of objects, each with one property/value pair
+ * @returns {array[]}          Array of arrays of objects grouped by property
+ */
+function groupByProperty(objects) {
+    var groupedObj = objects.reduce(function(accumulator, obj) {
+        var prop = Object.keys(obj)[0];
+        accumulator[prop] = accumulator[prop] ? accumulator[prop].concat(obj) : [obj];
+        return accumulator;
+    }, {});
+    return Object.keys(groupedObj).map(function(prop) {
+        return groupedObj[prop];
+    });
+}
+
+/**
+ * Create valid rule configurations by combining two arrays,
+ * with each array containing multiple objects each with a
+ * single property/value pair and matching properties.
+ *
+ * e.g.:
+ * combinePropertyObjects(
+ *     [{before: true}, {before: false}],
+ *     [{after: true}, {after: false}]
+ * );
+ *
+ * will return:
+ * [
+ *     {before: true, after: true},
+ *     {before: true, after: false},
+ *     {before: false, after: true},
+ *     {before: false, after: false}
+ * ]
+ *
+ * @param   {Object[]} objArr1 Single key/value objects, all with the same key
+ * @param   {Object[]} objArr2 Single key/value objects, all with another key
+ * @returns {Object[]}         Combined objects for each combination of input properties and values
+ */
+function combinePropertyObjects(objArr1, objArr2) {
+    var res = [];
+    if (objArr1.length === 0) {
+        return objArr2;
+    }
+    if (objArr2.length === 0) {
+        return objArr1;
+    }
+    objArr1.forEach(function(obj1) {
+        objArr2.forEach(function(obj2) {
+            var combinedObj = {};
+            var obj1Props = Object.keys(obj1);
+            var obj2Props = Object.keys(obj2);
+            obj1Props.forEach(function(prop1) {
+                combinedObj[prop1] = obj1[prop1];
+            });
+            obj2Props.forEach(function(prop2) {
+                combinedObj[prop2] = obj2[prop2];
+            });
+            res.push(combinedObj);
+        });
+    });
+    return res;
+}
+
+
+/**
+ * Creates a new instance of a rule configuration set
```

Comment:
Yes, I will add to the JSDoc, but it is an array of configurations that are valid for a given rule.  For example, the configurations for the `semi` rule could be:

```js
ruleConfigSet.ruleConfigs // -> [[2], [2, "always"], [2, "never"]]
```